File: src/main/java/edu/dcccd/trans/service/UserDetailsServiceImpl.java
package edu.dcccd.trans.service;

import edu.dcccd.trans.entity.Role;
import edu.dcccd.trans.entity.User;
import edu.dcccd.trans.repository.RoleRepository;
import edu.dcccd.trans.repository.UserRepository;
import org.springframework.beans.factory.annotation.Autowired;

import org.springframework.security.core.GrantedAuthority;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UserDetailsService;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.List;

@Service
public class UserDetailsServiceImpl implements UserDetailsService {

	@Autowired
	private UserRepository userRepository;
	@Autowired
	private RoleRepository roleRepository;

    @Override
    public UserDetails loadUserByUsername(String userName) throws UsernameNotFoundException {
        User user = userRepository.findByUsername(userName);

        if (user == null) {
            throw new UsernameNotFoundException("User " + userName + " was not found in the database");
        }

        // [ROLE_USER, ROLE_ADMIN,..]
        List<Role> roleNames = roleRepository.findByUserId(user.getId());

        List<GrantedAuthority> grantList = new ArrayList<>();
        if (roleNames != null) {
            for (Role role : roleNames) {
                // ROLE_USER, ROLE_ADMIN,..
                GrantedAuthority authority = new SimpleGrantedAuthority(role.getRole());
                grantList.add(authority);
            }
        }

        return new org.springframework.security.core.userdetails.User(user.getUsername(),
                user.getEncryptedPassword(), grantList);
    }
}

File: src/main/kotlin/edu/dcccd/trans/service/UserDetailsServiceImpl.kt
package edu.dcccd.trans.service

import edu.dcccd.trans.repository.RoleRepository
import edu.dcccd.trans.repository.UserRepository
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.security.core.GrantedAuthority
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.security.core.userdetails.UsernameNotFoundException
import org.springframework.stereotype.Service
import java.util.*

@Service
class UserDetailsServiceImpl : UserDetailsService {

    @Autowired
    private val userRepository: UserRepository? = null
    @Autowired
    private val roleRepository: RoleRepository? = null

    @Throws(UsernameNotFoundException::class)
    override fun loadUserByUsername(userName: String): UserDetails {
        val (id, username, encryptedPassword) = userRepository!!.findByUsername(userName)
                ?: throw UsernameNotFoundException("User $userName was not found in the database")

// [ROLE_USER, ROLE_ADMIN,..]
        val roleNames = roleRepository!!.findByUserId(id)

        val grantList = ArrayList<GrantedAuthority>()
        if (roleNames != null) {
            for (role in roleNames) {
                // ROLE_USER, ROLE_ADMIN,..
                val authority = SimpleGrantedAuthority(role.role)
                grantList.add(authority)
            }
        }

        return org.springframework.security.core.userdetails.User(username,
                encryptedPassword, grantList)
    }
}


